src/particle.ts

```diff
@@ -0,0 +1,98 @@
+import {Random} from "@reside-ic/random";
+
+import type {DustModel, DustModelConstructable, DustModelInfo} from "./model";
+
+export class Particle {
+    /** The step (in time) that the particle is currently at */
+    public step: number;
+
+    /** The size of the particle state - this cannot be modified after
+     * creation
+     */
+    public readonly size: number;
+
+    private _model: DustModel;
+    private _y: number[];
+    private _yNext: number[];
+
+    /**
+     * Construct a particle
+     *
+     * @param model An initialised model - note that unlike the C++
+     * version of dust, all particles share the same initialised model
+     * as they do not run in parallel.
+     *
+     * @param step The step to start the particle at
+     *
+     * @param state Optional vector of state; if given then this is
+     * the initial state for the particle. Otherwise the state comes
+     * from the models `initial` method (see {@link DustModel})
+     */
+    constructor(model: DustModel, step: number, state?: readonly number[]) {
+        this._model = model;
+        this.step = step;
+        this.size = model.size();
+        if (state === undefined) {
+            this._y = model.initial(this.step);
+        } else {
+            this._y = [...state];
+        }
+        this._yNext = Array(this.size).fill(0);
+    }
+
+    /**
+     * Model information, about ordering of variables, parameters,
+     * etc.
+     */
+    public info(): DustModelInfo {
+        return this._model.info();
+    }
+
+    /**
+     * Run a particle up to the end of `stepEnd`
+     *
+     * @param stepEnd Step to run the particle to the end of
+     *
+     * @param random The random number state to use for any stochastic
+     * draws
+     */
+    public run(stepEnd: number, random: Random): void {
+        while (this.step < stepEnd) {
+            this._model.update(this.step++, this._y, this._yNext, random);
+            this.swap();
+        }
+    }
+
+    /**
+     * Return current particle state
+     *
+     * @param index May be removed?
+     */
+    public state(index: number[] | null): readonly number[] {
+        if (index === null) {
+            return this._y;
+        } else {
+            // TODO, this path is now never used
+            return index.map((i: number) => this._y[i]);
```

Comment:
confused by the TODO, this path is referenced in the tests at least? https://github.com/mrc-ide/dust-js/pull/1/files#diff-7ffef65bb1ef225b67b4e49e888c5374b565160a6ab4519fd1330c6f68b91197R25

Reply:
Sorry, this PR is derived from a bunch of proof-of-concept work I did, then fitted into this (and subsequent) PRs. This was the interface for getting some or all of the state, but it's not actually that useful here it turns out with just one language to play with. I've removed this in 6dffcb8 now